### PR TITLE
Us076 - Amend Variable event dates

### PR DESCRIPTION
**Pull request by jwgwalton**
Validation for editing non mandatory event dates

### CODE REVIEW

**Comment by jwgwalton** [comment]
[US076: Amend Variable Event Dates []](https://trello.com/c/OnoJESML/158-us076-amend-variable-event-dates)

**Comment by jwgwalton** [comment]
Shouldn't go in until the design work on scheduling has been done.

File: src/main/java/uk/gov/ons/ctp/response/collection/exercise/service/impl/EventValidator.java [hunk 1]
```diff
@@ -13,41 +14,97 @@
...
      * @param existingEvents
      * @param updatedEvent
      * @return
      */
-    public boolean validate(final List<Event> existingEvents, final Event updatedEvent) {
+    public boolean validate(final List<Event> existingEvents, final Event updatedEvent,
+                            final CollectionExerciseDTO.CollectionExerciseState collectionExerciseState) {
+
+        // Can only update reminders of the non mandatory events when READY_FOR_LIVE
+        if (collectionExerciseState.equals(CollectionExerciseDTO.CollectionExerciseState.READY_FOR_LIVE)
```
Comment (tej99) [inline src/main/java/uk/gov/ons/ctp/response/collection/exercise/service/impl/EventValidator.java]: Should the check be for READY_FOR_LIVE state and LIVE states?

File: src/main/java/uk/gov/ons/ctp/response/collection/exercise/service/impl/EventValidator.java [hunk 2]
```diff
@@ -13,41 +14,97 @@
...
+                && (!isReminder(updatedEvent) && !isMandatory(updatedEvent))) {
```
Comment (tej99) [inline src/main/java/uk/gov/ons/ctp/response/collection/exercise/service/impl/EventValidator.java]: I thought a reminder can be set even after the CE is READY_FOR_LIVE/ LIVE.

File: src/main/java/uk/gov/ons/ctp/response/collection/exercise/service/impl/EventValidator.java [hunk 3]
```diff
@@ -13,41 +14,97 @@
...
+            return false;
+        }
 
         Optional<Event> existingEvent = existingEvents.stream().findFirst().filter(
                 event -> event.getTag().equals(updatedEvent.getTag()));
 
-        if (isEventInPast(existingEvent)) {
+        // Mandatory & Reminder events can't be updated if already past
+        if ((isMandatory(updatedEvent) || isReminder(updatedEvent)) && isEventInPast(existingEvent)) {
```
Comment (tej99) [inline src/main/java/uk/gov/ons/ctp/response/collection/exercise/service/impl/EventValidator.java]: Is this not the case for all events not just the mandatory and reminder events?

File: src/main/java/uk/gov/ons/ctp/response/collection/exercise/service/impl/EventValidator.java [hunk 4]
```diff
@@ -13,41 +14,97 @@
...
+     */
+    private boolean validateMandatoryEvents(final Map<String, Event> eventMap) {
+        Event mpsEvent = eventMap.get(EventService.Tag.mps.toString());
+        Event goLiveEvent = eventMap.get(EventService.Tag.go_live.toString());
+        Event returnByEvent = eventMap.get(EventService.Tag.return_by.toString());
+        Event exerciseEndEvent = eventMap.get(EventService.Tag.exercise_end.toString());
+
+        return mpsEvent.getTimestamp().before(goLiveEvent.getTimestamp())
+                && goLiveEvent.getTimestamp().before(returnByEvent.getTimestamp())
+                && returnByEvent.getTimestamp().before(exerciseEndEvent.getTimestamp());
```
Comment (tej99) [inline src/main/java/uk/gov/ons/ctp/response/collection/exercise/service/impl/EventValidator.java]: The check before was !returnByEvent.getTimestamp().before(exerciseEndEvent.getTimestamp())

Comment (jwgwalton) [reply src/main/java/uk/gov/ons/ctp/response/collection/exercise/service/impl/EventValidator.java]: it had return false after that, so if the return byevent wasn't before exercise end it was invalid.

File: src/main/java/uk/gov/ons/ctp/response/collection/exercise/service/impl/EventValidator.java (hunk 3 above)
Comment (jwgwalton) [reply src/main/java/uk/gov/ons/ctp/response/collection/exercise/service/impl/EventValidator.java]: This follows the validation on the trello card.
https://digitaleq.atlassian.net/wiki/spaces/RASB/pages/259555593/US076+INT+Amend+Variable+Event+Dates

File: src/main/java/uk/gov/ons/ctp/response/collection/exercise/service/impl/EventValidator.java (hunk 2 above)
Comment (jwgwalton) [reply src/main/java/uk/gov/ons/ctp/response/collection/exercise/service/impl/EventValidator.java]: Nope https://digitaleq.atlassian.net/wiki/spaces/RASB/pages/259555593/US076+INT+Amend+Variable+Event+Dates

Comment (tej99) [reply src/main/java/uk/gov/ons/ctp/response/collection/exercise/service/impl/EventValidator.java]: When the collection exercise has been set 'Ready for Live', the user is only able to amend the reminder dates

Given: The internal user has set the collection exercise as 'Ready for Live'
When: They navigate to the collection exercise details page
Then: They are able to edit the reminder dates (only)

File: src/main/java/uk/gov/ons/ctp/response/collection/exercise/service/impl/EventValidator.java (hunk 3 above)
Comment (tej99) [reply src/main/java/uk/gov/ons/ctp/response/collection/exercise/service/impl/EventValidator.java]: "All event dates cannot be changed if they have past"

File: src/main/java/uk/gov/ons/ctp/response/collection/exercise/service/impl/EventValidator.java (hunk 4 above)
Comment (tej99) [reply src/main/java/uk/gov/ons/ctp/response/collection/exercise/service/impl/EventValidator.java]: ah okay

**Comment by jwgwalton** [comment]
NOT TO BE MERGED, THIS CHANGE DOESN'T AFFECT THE ACTUAL DATES WHICH GOVERN THE SYSTEM.

**Pull request merged**